### DISCUSSION TITLE
Support for HLK-LD2412

### DISCUSSION THREAD

**Discussion by pkkrusty** [arendst/Tasmota#21976] [Ideas]
LD2412 is an improved version of LD2410 with better range and wider FOV. Similar price point. I have a few on order and will see how compatible they are with the LD2410 driver. 

But wanted to preemptively ask for the LD2410 driver to add support for 2412. (I know PRs are welcome from anyone, but Theo wrote the code for 2410...)

Esphome community is actively working on support [link](https://community.home-assistant.io/t/diy-human-sensor-l12-base-on-hlk-ld2412/758421)

**Comment by Staars** [comment]
Nice finding. I received one too now and started to adapt the old BLE driver of the LD2410.

It is a little bit different now, but finally I got notifications from the sensor:
```
18:23:56.872 BLE: start LD2412 driver
18:23:56.874 BLE: Connection Ctx created
18:23:56.876 BLE: discover services
18:23:58.773 LD2: Start config mode
18:23:58.919 LD2: Sensor ACK
18:23:59.773 LD2: Request FW
18:24:00.119 LD2: Firmware version V1.36.36080523
18:24:00.774 LD2: Stop config mode
18:24:00.870 bytes('FDFCFBFA0400FE01000004030201')
```

The data format has changed according to the new number of gates, but most of the commands are still there and working.

**Reply by DarthWeber** [reply]
Happy to read that. So is it possible now to use the LD2412 with tasmota?

**Reply by sfromis** [reply]
Well, "started to adapt" does not mean to expect it to be completed in 12 hours after the post...

**Reply by DarthWeber** [reply]
> Well, "started to adapt" does not mean to expect it to be completed in 12 hours after the post...

That's not what i mean or expect. I mean:

> most of the commands are still there and working.

 if it is possibnle to partially use the LD2412 with the existing driver for the LD2410 already.

**Reply by sfromis** [reply]
My point was that being able to parse and process sensor data is more important than the set of commands, but when "data format has changed"...

**Reply by ortegafernando** [reply]
@Staars @sfromis Thank for all your effort guys. Do you think it will be possible to add also HLK-LD2410S (less than 1mA version) ?

**Reply by Staars** [reply]
I have worked my way through the various documents, that can be found on HLK's website and in some GitHub repos.

In theory it would be possible to write one (Berry) driver, that can work with most sensors of the LD24-family, but there are a lot of "if's" and "but's".

There are definitely missing commands in the official documentation, that would have to be reverse engineerd (that is an unnessecary waste of time!)

The received payload in the case of the LD2412 does not match the documentation completely. That could be attributed to the fact, that I use BLE and not serial, but it could also mean, that changes in the sensor firmware are responsible for that. In the case of the LD2410 this did definitely happen.

AFAIK there is no API to simply get a sensor/hardware ID, so that I know if it is a LD2410/11/12 or another type. This would have been really nice for a universal driver approach. It might be possible to infer this via the firmware info or something like that, but again - a waste of development time.

So summarized: I am almost sure, that it is technically possible to support "everything", but at the moment it looks like an annoying task due to the lack of documentation. I have my doubts, that other drivers out there, will work with every firmware version (mainly of the very common LD2410) as supposed.

**Reply by Staars** [reply]
Just to answer myself to some extent...

I started a second driver in Berry for serial communication and can confirm for the LD2412, that the payload seems to be equal for BLE and serial. Thus in my my case both do not match the documentation (in engineering mode).

But besides the idea is growing in me to try to have a common LD24xx driver, that infers the connected sensor via firmware type and firmware version (this command should be supported by every sensor of the family) and then instantiates the fitting sensor class, that has to implement some mandatory functions.

**Reply by Staars** [reply]
https://github.com/Staars/berry-examples/blob/main/serial/serial_ld24xx.be
This is just a proof of concept, it is not even in alpha state.

If someone wants to give it a try there is basically only one question from me at the moment. Does the driver detect the correct sensor type?

`var ld2 = LD2(3,4,115200) # or 256000 baud` in line 639 must be changed according to the pin setup `LD2(rapine,txpin,baud)`.

**Reply by Staars** [reply]
As I have the LD2420 connected to my test ESP, I did continue with this radar sensor.
The ESPHome driver is pretty well documented and the resulting data is in line with what I receive here.

I updated the driver from above and it works here for me with auto detection of the LD2420, outputting the full gate energies (like the engineering mode on the other sensors):
`"LD2420":{"presence":1,"range":134,"gates":[11817,2900,185,68,18,41,32,25,17,20,10,13,10,13,13,0]}`

I will try my luck with the LD2412 next, but I expect problems with unknown/not documented parts of the payload with the firmware version of my LD2412.

**Reply by Staars** [reply]
Finally most of the docs make sense to me.
I think, I have the engineering mode working - producing a JSON like that:
`"LD2412":{"distance":[0,40],"energy":[0,100],"moving_energies":[0, 2, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0],"static_energies":[0, 100, 100, 100, 100, 100, 100, 100, 100, 13, 78, 70, 64],"light":11}`.

<img width="538" alt="ld2412_ble" src="https://github.com/user-attachments/assets/04fcce85-de7f-4c8b-bf54-b8551d88d9a3">
(LD2412 connected via Bluetooth)

**Reply by pkkrusty** [reply]
Current status of this project? Asking because of recent work adapting the 2410 driver for 2402 model (which is incredibly cheap). Is it possible to use 2412 with serial comms at the moment? Or just BLE?

**Reply by Staars** [reply]
The driver is unchanged since then and did work for me. It was possible to connect any of my LD2410, LD2412 or LD2420 via serial and the rest was more or less plug and play, means correct autodetection of the particular model.
I will probably order a LD2402 too and check, if it is possble to integrate it into the generic driver, which I personally prefer over a fragmented approach.

The whole LD24 family is not really designed for auto detection (of features) and a firmware upgrade can ruin the party, but it would probably possible to fix upcoming problems with specialized adaptions for different firmware versions in the model specific part of the driver.

**Comment by jzhvymetal** [comment]
I'm trying to use this script.  I got to start communicate but get the following below.  Any idea?


17:58:50.681 CMD: Br load('serial_ld24xx.be')
17:58:50.831 LD2: start LD24xx driver
17:58:50.835 bytes('FDFCFBFA0400FF00010004030201')
17:58:50.838 RSL: RESULT = {"Br":"true"}
17:58:50.859 bytes('FDFCFBFA0200A00004030201')
17:58:50.859 LD2: Request FW, try to identify sensor type
17:58:50.936 LD2: FW type 100
17:58:50.937 LD2: found LD2410 2.68.12560
17:58:50.939 bytes('FDFCFBFA0200000004030201')
17:58:50.940 LD2: Request Version, try to identify sensor type
17:58:50.987 BRY: Exception> 'index_error' - list index out of range
17:58:51.036 BRY: Exception> 'index_error' - list index out of range
17:58:51.087 BRY: Exception> 'index_error' - list index out of range
17:58:51.136 BRY: Exception> 'index_error' - list index out of range
17:58:51.186 BRY: Exception> 'index_error' - list index out of range
17:58:51.236 BRY: Exception> 'index_error' - list index out of range

**Reply by Staars** [reply]
This is probably an unexpected response for the driver, maybe due to a different firmware version, which might produce a payload that then gets parsed incorrectly.
I will check with my sensor, but this will take a little bit of time.

**Reply by jzhvymetal** [reply]
Something is not parsing correctly  I tried downgrading firmware and get the following:

LD2410  Firmware 2.04.23022511 ->found LD2410 2.4.9489
LD2410  Firmware 2.44.24073110 ->found LD2410 2.68.12560

Can not try other firmware's as these are the only that are in the Phone app via Bluetooth.  Also since I have using TX/RX pin via software I also tried lowering baud rate 57600

**Reply by Staars** [reply]
I am a bit late, but:

17:58:50.939 bytes('FDFCFBFA0200000004030201')
is a NACK response, that may not have happened with the firmware I have here for testing.

I just updated the driver to return early on NACK. Maybe that helps to go further. There are "good" chances, that we face other problems later with different payload formats.

The firmware name parsing does not really look like a problem to me, because this is very likely only the difference in representig values as decimal, hex or ascii. But I get the point, that it would be less confusing, if I do it like the BLE app does.

If you like, you can test the updated driver.
